rust-runtime/smithy-http/src/event_stream.rs

```diff
@@ -0,0 +1,303 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+//! Provides Sender/Receiver implementations for Event Stream codegen.
+
+use crate::body::SdkBody;
+use crate::result::SdkError;
+use bytes::Bytes;
+use bytes_utils::SegmentedBuf;
+use futures_core::Stream;
+use hyper::body::HttpBody;
+use pin_project::pin_project;
+use smithy_eventstream::frame::{
+    DecodedFrame, MarshallMessage, MessageFrameDecoder, SignMessage, UnmarshallMessage,
+};
+use std::error::Error as StdError;
+use std::marker::PhantomData;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+/// Adapts a `Stream<SmithyMessageType>` to a signed `Stream<Bytes>` by using the provided
+/// message marshaller and signer implementations.
+///
+/// This will yield an `Err(SdkError::ConstructionFailure)` if a message can't be
+/// marshalled into an Event Stream frame, (e.g., if the message payload was too large).
+#[pin_project]
+pub struct MessageStreamAdapter<T, E> {
+    marshaller: Arc<dyn MarshallMessage<Input = T> + Send + Sync>,
```

Comment:
why is this an Arc here and a Box elsewhere?

Reply:
I think the last commit fixed this.